contrib(inventory): adds a simple dynamic inventory for Scaleway cloud provider

paulRbr opened the pull request:
##### SUMMARY

This PR adds a dynamic inventory to target inventories of Scaleway cloud provider. Adding servers into all the following groups:

```
[hostname1]
<public_ip|private_ip>

[hostname2]
<public_ip|private_ip>

[hostname3]
<public_ip|private_ip>

[all:children]
hostname1
hostname2

[tag1:children]
hostname1
hostname3

[tag2:children]
hostname2
hostname3

[region1:children]
hostname1

[region2:children]
hostname2
```
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME

`contrib/inventory`

paulRbr committed:
contrib(inventory): adds a dynamic inventory for Scaleway cloud
provider

Provides a dynamic inventory plugin for Scaleway cloud provider with
the following features:

- Configurable scaleway.ini file
- Cache API responses
- Choose public or private IPs
- Create groups per Scaleway 'tags'
- Create groups per Scaleway regions

paulRbr commented:
Hi @jborean93 would you be able to help to get feedback / review for this new feature? Thanks!

jborean93 commented:
Hey @paulRbr I wouldn't really be able to help with this as I don't have experience with the Scaleway product itself. Your best bet is to ask around on the IRC channel's (#ansible-devel) to see if someone can help with this.

nerzhul commented:
hello @paulRbr can you call it scaleway without the underscore please ?
Also you miss the region attribute. 
I think it's better to implement it without using the python sdk to remove the dependency, as the call is relatively simple to implement the call without using any dependency.
Did you get time this week to refactor it permitting to add it to 2.5 before the freeze ?

https://developer.scaleway.com/#servers-servers-get

paulRbr commented:
Hi @nerzhul, thanks for looking into this.

I have rewritten the plugin to not use the Scaleway SDK and use the `requests` package instead. Do you think `requests` is a "common" enough dependency?

Also I have added the region attribute as a configuration option or reading from `env` and creating groups in the inventory for each region.

Let me know what you think.

nerzhul commented on contrib/inventory/scaleway.py:
```diff
@@ -0,0 +1,223 @@
+
+    # Read config
+    config = ConfigParser.SafeConfigParser()
+    for configfilename in [os.path.abspath(sys.argv[0]).rsplit('.py')[0] + '.ini', 'scaleway.ini']:
```
just call scaleway.ini we are in ansible conflict should not occur

paulRbr replied:
@nerzhul I usually use dynamic inventories together with static inventories. As such all my inventories are in a directory, i.e:
```
├── integ/
│   ├── scaleway.ini
│   ├── scaleway.py
│   └── hosts
└── prod/
     ├── ec2.ini
     ├── ec2.py
     ├── scaleway.ini
     ├── scaleway.py
     └── hosts
```
When I call ansible it is from the root directory as such: `ansible-playbook -i prod my-playbook.yml`. This is why I need to read the `*.ini*` file from the running script.

(EC2 inventory does exactly this here: https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L308-L309)

nerzhul commented:
Except the minor change, it sounds nice. requests is better yes, as we are only using one call using a new dep is a little bit heavy :)

nerzhul commented:
bot_status

paulRbr commented:
@nerzhul I have replied to your comment and think we need to keep the relative path for loading the inventory config :)

nerzhul commented:
it's an interesting argument

nerzhul commented:
shipit

nerzhul commented:
tested works well

```
{"scw-xxxxd2": ["10.77.88.3"], "all": {"hosts": [], "children": ["scw-xxxxd2"]}, "par1": {"hosts": [], "children": ["scw-xxxxd2"]}}%
```

mscherer commented on contrib/inventory/scaleway.py:
```diff
@@ -0,0 +1,223 @@
+        try:
+            existing = os.stat(
+                '/'.join([dpath, 'scaleway_ansible_inventory.json']))
+        except:
```
I tend to dislike unlimited except, since this can hide problem that shouldn't be hidden. Can this be limited somehow to just catch a few exception ?

paulRbr replied:
You are right. I have fixed that now.

mscherer commented on contrib/inventory/scaleway.ini:
```diff
@@ -0,0 +1,34 @@
+#
+[cache]
+cache_max_age = 60
+cache_dir = /tmp
```
That would result into insecure default, since someone would be able to inject a cache file in a shared server. This would also fail if there is concurent run of ansible using the same inventory.

paulRbr replied:
Indeed. I've fixed that, thanks!

mscherer commented on contrib/inventory/scaleway.py:
```diff
@@ -0,0 +1,223 @@
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+import ConfigParser
```
This is not python3 compliant.

paulRbr replied:
Thanks, i've shamelessly copied from the ec2 inventory: https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.py#L313-L316

nerzhul commented:
i tested in both python 2 and python 3 is works nice

nerzhul commented:
shipit

mscherer commented:
shipit

nerzhul commented:
bot_status

paulRbr commented:
ready_for_review

nerzhul commented:
shipit

paulRbr commented:
Thanks @nerzhul @mscherer

nerzhul commented:
thanks you for your speed permitting to merge this for 2.5, freeze is in 2 days :)

harobed commented:
@paulRbr thanks for your contrib 👍 

For now, I use [bearstech/scw_ansible](https://github.com/bearstech/scw_ansible).
With this tool I can use variable env to store my Scaleway Token.

What do you think about add variable env support in your script ?

paulRbr commented:
Hi @harobed!

The current dynamic inventory script reads the `SCALEWAY_TOKEN` env
variable before reading the configuration file. Did you try it out?

Let me know if that is what you needed. I guess we could add more
explanation in the documentation header to make it clear.

Hope it helps!

Em 02/02/2018 18:28, "Stéphane Klein" <notifications@github.com> escreveu:

> @paulRbr <https://github.com/paulrbr> thanks for your contrib 👍
>
> For now, I use bearstech/scw_ansible
> <https://github.com/bearstech/scw_ansible>.
> With this tool I can use variable env to store my Scaleway Token.
>
> What do you think about add variable env support in your script ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/30313#issuecomment-362633501>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3MAfq3NZ3j-JzzSK0L45BMWgrnisL1ks5tQzdOgaJpZM4PWzKn>
> .
>

harobed commented:
@paulRbr 

> The current dynamic inventory script reads the `SCALEWAY_TOKEN` env variable before reading the configuration file. Did you try it out? 

I didn't see it, ok I'll try it, thanks.

> I guess we could add more explanation in the documentation header to make it clear.

👍
